### PR TITLE
Static Files to .NET 6 /4

### LINKED ISSUE #24094: .NET 6: Static files
Also do #18525, just mention
 use MVC's path pattern matching to limit the redundancy.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d47f38e2-bc36-cfb8-b0cc-17a534689f8b
* Version Independent ID: 3fec6e08-fc99-7a5c-796f-3f2347cad891
* Content: [Static files in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/static-files?view=aspnetcore-6.0)
* Content Source: [aspnetcore/fundamentals/static-files.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/static-files.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes  #24094
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/static-files?view=aspnetcore-6.0&branch=pr-en-us-24166)

### CODE REVIEW

File: aspnetcore/fundamentals/static-files.md [hunk 1]
```diff
@@ -101,45 +91,43 @@ To serve static files based on authorization:
...
 
 Directory browsing allows directory listing within specified directories.
 
 Directory browsing is disabled by default for security reasons. For more information, see [Security considerations for static files](#security-considerations-for-static-files).
 
-Enable directory browsing with:
-
-* <xref:Microsoft.Extensions.DependencyInjection.DirectoryBrowserServiceExtensions.AddDirectoryBrowser%2A> in `Startup.ConfigureServices`.
-* <xref:Microsoft.AspNetCore.Builder.DirectoryBrowserExtensions.UseDirectoryBrowser%2A> in `Startup.Configure`.
+Enable directory browsing with <xref:Microsoft.AspNetCore.Builder.DirectoryBrowserExtensions.UseDirectoryBrowser%2A>:
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/static-files.md:115]: @Tratcher  what is `AddDirectoryBrowser` used for? I don't need it to browse directories. I guess this document has always been wrong.  `UseDirectoryBrowser` is all you need.

Comment (Tratcher) [reply aspnetcore/fundamentals/static-files.md:115]: DirectoryBrowserMiddlware needs some services and those are added by AddDirectoryBrowser.
https://github.com/dotnet/aspnetcore/blob/fc4e391aa58a9fa67fdc3a96da6cfcadd0648b17/src/Middleware/StaticFiles/src/DirectoryBrowserServiceExtensions.cs#L25
https://github.com/dotnet/aspnetcore/blob/fc4e391aa58a9fa67fdc3a96da6cfcadd0648b17/src/Middleware/StaticFiles/src/DirectoryBrowserMiddleware.cs#L46

You may get lucky if those services were already added by some other component like MVC. Keep AddDirectoryBrowser in the sample anyways.

File: aspnetcore/fundamentals/static-files.md [hunk 2]
```diff
@@ -40,17 +36,17 @@ Consider creating the *wwwroot/images* folder and adding the *wwwroot/images/MyI
...
+[!code-csharp[](~/fundamentals/static-files/samples/6.x/StaticFilesSample/Program.cs?name=snippet&highlight=15)]
 
 The parameterless `UseStaticFiles` method overload marks the files in [web root](xref:fundamentals/index#web-root) as servable. The following markup references *wwwroot/images/MyImage.jpg*:
 
 ```html
 <img src="~/images/MyImage.jpg" class="img" alt="My image" />
 ```
 
-In the preceding code, the tilde character `~/` points to the [web root](xref:fundamentals/index#web-root).
+In the preceding highlighted markup, the tilde character `~` points to the [web root](xref:fundamentals/index#web-root).
```
Comment (serpent5) [inline aspnetcore/fundamentals/static-files.md:49]: There's no highlight now as this is just one line.

```suggestion
In the preceding markup, the tilde character `~` points to the [web root](xref:fundamentals/index#web-root).
```

File: aspnetcore/fundamentals/static-files.md [hunk 3]
```diff
@@ -66,27 +62,23 @@ Consider a directory hierarchy in which the static files to be served reside out
...
 In the preceding code, the *MyStaticFiles* directory hierarchy is exposed publicly via the *StaticFiles* URI segment. A request to `https://<hostname>/StaticFiles/images/red-rose.jpg` serves the *red-rose.jpg* file.
 
-The following markup references *MyStaticFiles/images/red-rose.jpg*:
-
-```html
-<img src="~/StaticFiles/images/red-rose.jpg" class="img" alt="A red rose" />
-```
+The following highlighted markup references *MyStaticFiles/images/red-rose.jpg*:
+<!-- zz test via /Home2/MyStaticFilesRR -->
+[!code-html[](~/fundamentals/static-files/samples/6.x/StaticFilesSample/Views/Home2/MyStaticFilesRR.cshtml?range=1&highlight=1)]
```
Comment (serpent5) [inline aspnetcore/fundamentals/static-files.md:71]: This is just a single line, so I suggest removing the highlight.

```suggestion
The following markup references *MyStaticFiles/images/red-rose.jpg*:
<!-- zz test via /Home2/MyStaticFilesRR -->
[!code-html[](~/fundamentals/static-files/samples/6.x/StaticFilesSample/Views/Home2/MyStaticFilesRR.cshtml?range=1)]
```

File: aspnetcore/fundamentals/static-files.md [hunk 4]
```diff
@@ -101,45 +93,45 @@ To serve static files based on authorization:
...
-
-[!code-csharp[](~/fundamentals/static-files/samples/3.x/StaticFilesSample/StartupBrowse.cs?name=snippet_ClassMembers&highlight=4,21-35)]
+[!code-csharp[](~/fundamentals/static-files/samples/6.x/StaticFilesSample/Program.cs?name=snippet_db&highlight=8,22-27)]  
 
+<!-- Select RP Home > Directory browsing -->
 The preceding code allows directory browsing of the *wwwroot/images* folder using the URL `https://<hostname>/MyImages`, with links to each file and folder:
 
 ![directory browsing](static-files/_static/dir-browse.png)
 
+For basic programs, `UseDirectoryBrowser` without `AddDirectoryBrowser` may work. [AddDirectoryBrowser](https://github.com/dotnet/aspnetcore/blob/fc4e391aa58a9fa67fdc3a96da6cfcadd0648b17/src/Middleware/StaticFiles/src/DirectoryBrowserServiceExtensions.cs#L25) should be called as it [adds services](https://github.com/dotnet/aspnetcore/blob/fc4e391aa58a9fa67fdc3a96da6cfcadd0648b17/src/Middleware/StaticFiles/src/DirectoryBrowserMiddleware.cs#L46) like [HtmlEncoder](/dotnet/api/system.text.encodings.web.htmlencoder).
```
Comment (serpent5) [inline aspnetcore/fundamentals/static-files.md:126]:
```suggestion
`AddDirectoryBrowser` [adds services](https://github.com/dotnet/aspnetcore/blob/fc4e391aa58a9fa67fdc3a96da6cfcadd0648b17/src/Middleware/StaticFiles/src/DirectoryBrowserServiceExtensions.cs#L25) required by the directory browsing middleware, including <xref:System.Text.Encodings.Web.HtmlEncoder>. These services may be added by other calls, such as <xref:Microsoft.Extensions.DependencyInjection.MvcServiceCollectionExtensions.AddRazorPages%2A>, but we recommend calling `AddDirectoryBrowser` to ensure the services are added in all apps.
```

File: aspnetcore/fundamentals/static-files/samples/6.x/StaticFilesSample/Program.cs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Error");
+    app.UseHsts();
+}
+
+app.UseHttpsRedirection();
+
+// Set up custom content types - associating file extension to MIME type
+var provider = new FileExtensionContentTypeProvider();
```
Comment (Tratcher) [inline aspnetcore/fundamentals/static-files/samples/6.x/StaticFilesSample/Program.cs:303]: This provider needs to be assigned to the StaticFileOptions.ContentTypeProvider property.

**Pull request merged**